[Plugin] Allow selectively disabling postprocessing

TurboTurtle opened the pull request:
Adds two mechanisms by which users can choose to disable postprocessing
of collected information.

First, is a global method exposed via the `--no-postproc` option. Using
this option will skip postprocessing for all plugins.

Second, is a per-plugin option exposed via a new 'postproc' plugin
option. This is set to _True_ by default (meaning yes, perform
postprocessing), which users can set to False or off to disable
postprocessing for that plugin only; e.g. `-k podman.postproc=off`

Closes: #286
Resolves: #1862

Signed-off-by: Jake Hunsaker <jhunsake@redhat.com>

---
Please place an 'X' inside each '[]' to confirm you adhere to our [Contributor Guidelines](https://github.com/sosreport/sos/wiki/Contribution-Guidelines)

- [x] Is the commit message split over multiple lines and hard-wrapped at 72 characters?
- [x] Is the subject and message clear and concise?
- [x] Does the subject start with **[plugin_name]** if submitting a plugin patch or a **[section_name]** if part of the core sosreport code?
- [x] Does the commit contain a **Signed-off-by: First Lastname <email@example.com>**?
- [x] If this commit closes an existing issue, is the line `Closes: #ISSUENUMBER` included in an independent line?
- [x] If this commit resolves an existing pull request, is the line `Resolves: #PRNUMBER` included in an independent line?

bmr-cymru commented on sos/plugins/__init__.py:
```diff
@@ -1570,6 +1574,15 @@ def setup_verify(self):
         if verify_cmd:
             self.add_cmd_output(verify_cmd)
 
+    def _postproc(self):
```
The scoping looks funny here - normally the public (non `_`-prefixed) version of the method is the wrapper, and the private method is the wrapped one.

I think it would be less confusing to re-name the existing method to `_postproc()`, and have `sosreport` continue to call `plug.postproc()` - then depending on the option value the public method calls or doesn't call `self._postproc()` to do the actual work.

TurboTurtle replied:
The plugins doing postprocessing today all override `postproc()`, hence why I went with this approach. So if `sosreport` still calls that directly, we'd need to change the plugins to now override `_postproc()` instead, no? Seemed a bit awkward to me.

The other option could be to do this check directly in `sosreport`, e.g.:

```python
    def postproc(self):
        for plugname, plug in self.loaded_plugins:
            try:
                if plug.get_option('postproc'):
                    plug.postproc()
                else:
                    self.log_info('some skip message')
```

bmr-cymru replied:
Yeah, I see what you mean - although I think having the `get_option()` check in sosreport looks better - we aren't super strict about public vs. private in sos (and neither is Python ;-), but it looks weird to have a "private" method called by another module, that calls a similarly-named "public" method.

bmr-cymru commented:
> My only comment would be it might be nice to add these to the .travis ci for automatically making sure they continue to work error free.

It would also be good to add a unit test to verify that `Plugin` properly does/doesn't call `postproc()` based on the option values.

TurboTurtle commented:
I've updated the approach to check for the `postproc` plugin option in `sosreport` rather than within `Plugin`, which matches the behavior to check for `--no-postproc` globally.

I've added a single plugin test that checks to make sure the plugin default for `postproc` remains `True`.  I've not found a way with the current unit test structure to properly test the actual checks `sosreport` does, but at least this check is _something_. All ears for any suggestions on this front (including how to test this with Travis @BryanQuigley)

BryanQuigley commented:
I need to propose a better way for travis going forward but just a simple:
sudo ~/virtualenv/python$TRAVIS_PYTHON_VERSION/bin/python ./sosreport --batch --no-postproc  --config-file=sos.conf

Would be a basic "check" - again doesn't check that it actually works though.

BryanQuigley commented:
No need to add anything more to travis.   I'll add this easily in the script in #1896 

I think all conversations are resolved?  Need any more reviews?

pmoravec commented:
ACK to the PR, just wouldnt it be better to have the options with equivalent (and not opposite) meaning? I mean have global `--no-postproc` (defaults to False) and plugin specific `nopostproc` (defaults to False), OR have global `--postproc` (defaults to True) and plugin specific `postproc` (defaults to True). It might be confusing to use complementary options global/plugin-wise - but that can be my subjective opinion only.

TurboTurtle commented:
I went with the plugin option of `postproc` since using `-k plugin.postproc=off` seems more intuitive than ` -k plugin.nopostproc=on`, I.E. "turn this off" versus "turn this on to turn this off". Similarly the global "turn it all off with this toggle" seems easier to grok to me.
